src/main/scala/com/databricks/spark/xml/package.scala

```diff
@@ -19,10 +19,64 @@ import scala.collection.Map
 
 import org.apache.hadoop.io.compress.CompressionCodec
 
+import org.apache.spark.annotation.Experimental
 import org.apache.spark.sql._
-import com.databricks.spark.xml.util.XmlFile
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
+import org.apache.spark.sql.types.StructType
+
+import com.databricks.spark.xml.util.{InferSchema, XmlFile}
 
 package object xml {
+
+  private def withExpr(expr: Expression): Column = new Column(expr)
+
+  /**
+   * Infers the schema of XML documents as strings.
+   *
+   * @param ds Dataset of XML strings
+   * @return inferred schema for XML
+   */
+  @Experimental
+  def inferSchema(ds: Dataset[String]): StructType =
+    inferSchema(ds, Map.empty[String, String])
+
+  /**
+   * Infers the schema of XML documents as strings.
+   *
+   * @param ds Dataset of XML strings
+   * @param options additional XML parsing options
+   * @return inferred schema for XML
+   */
+  @Experimental
+  def inferSchema(ds: Dataset[String], options: Map[String, String]): StructType =
```

Comment:
There's `schema_of_json` and `schema_of_csv` in Apache spark. We can don't add both in this PR I believe. I can work on both separately.

Reply:
Do you mean this should be called `schema_of_xml`, at least to start?
I think we have to have such a method, or there's no way to get the schema to pass to `from_xml`.

Reply:
Yeah, I meant we should at least name it `schema_of_xml`.

Also, there have been many discussions about implementation details with multiple Spark committers ...

it ended up with implementing an expression that infers the schema from single record (single JSON or CSV), instead of going through all data.

If you don't mind, I will work on that. I should explain the history in the PR description I suspect.